src/analysis/normed_space/normed_group_hom.lean

```diff
@@ -0,0 +1,450 @@
+add_monoid_hom.mk' (λ g, add_monoid_hom.mk' (λ f, g.comp f)
+  (by { intros, ext, exact g.map_add _ _ }))
+  (by { intros, ext, simp only [comp_apply, pi.add_apply, function.comp_app,
+                                add_monoid_hom.add_apply, add_monoid_hom.coe_mk', coe_add] })
+
+@[simp] lemma comp_zero (f : normed_group_hom V₂ V₃) : f.comp (0 : normed_group_hom V₁ V₂) = 0 :=
+by { ext, exact f.map_zero' }
+
+@[simp] lemma zero_comp (f : normed_group_hom V₁ V₂) : (0 : normed_group_hom V₂ V₃).comp f = 0 :=
+by { ext, refl }
+
+end normed_group_hom
+
+namespace normed_group_hom
+
+/-!### Kernel -/
+section kernels
+
+variables {V V₁ V₂ V₃ : Type*}
+variables [normed_group V] [normed_group V₁] [normed_group V₂] [normed_group V₃]
+variables (f : normed_group_hom V₁ V₂) (g : normed_group_hom V₂ V₃)
+
+/-- The kernel of a bounded group homomorphism. Naturally endowed with a `normed_group` instance. -/
+def ker : add_subgroup V₁ := f.to_add_monoid_hom.ker
+
+lemma mem_ker (v : V₁) : v ∈ f.ker ↔ f v = 0 :=
+by { erw f.to_add_monoid_hom.mem_ker, refl }
+
+/-- The inclusion of the kernel, as bounded group homomorphism. -/
+@[simps] def ker.incl : normed_group_hom f.ker V₁ :=
+{ to_fun := (coe : f.ker → V₁),
+  map_zero' := add_subgroup.coe_zero _,
+  map_add' := λ v w, add_subgroup.coe_add _ _ _,
+  bound' := ⟨1, λ v, by { rw [one_mul], refl }⟩ }
+
+/-- Given a normed group hom `f : V₁ → V₂` satisfying `g.comp f = 0` for some `g : V₂ → V₃`,
+    the corestriction of `f` to the kernel of `g`. -/
+@[simps] def ker.lift (h : g.comp f = 0) :
+  normed_group_hom V₁ g.ker :=
+{ to_fun := λ v, ⟨f v, by { erw g.mem_ker, show (g.comp f) v = 0, rw h, refl }⟩,
+  map_zero' := by { simp only [map_zero], refl },
+  map_add' := λ v w, by { simp only [map_add], refl },
+  bound' := f.bound' }
+
+@[simp] lemma ker.incl_comp_lift (h : g.comp f = 0) :
+  (ker.incl g).comp (ker.lift f g h) = f :=
+by { ext, refl }
+
+end kernels
+
+/-! ### Range -/
+section range
+
+variables {V V₁ V₂ V₃ : Type*}
+variables [normed_group V] [normed_group V₁] [normed_group V₂] [normed_group V₃]
+variables (f : normed_group_hom V₁ V₂) (g : normed_group_hom V₂ V₃)
+
+/-- The image of a bounded group homomorphism. Naturally endowed with a `normed_group` instance. -/
+def range : add_subgroup V₂ := f.to_add_monoid_hom.range
+
+lemma mem_range (v : V₂) : v ∈ f.range ↔ ∃ w, f w = v :=
+by { rw [range, add_monoid_hom.mem_range], refl }
+
+end range
+
+variables {V W V₁ V₂ V₃ : Type*}
+variables [normed_group V] [normed_group W]
+variables [normed_group V₁] [normed_group V₂] [normed_group V₃]
+variables {f : normed_group_hom V W}
+
+/-- A `normed_group_hom` is *norm-nonincreasing* if `∥f v∥ ≤ ∥v∥` for all `v`. -/
+def norm_noninc (f : normed_group_hom V W) : Prop :=
+∀ v, ∥f v∥ ≤ ∥v∥
+
+/-- A isometry `normed_group_hom` is a `normed_group_hom` that preserves the norm. -/
+def is_isometry (f : normed_group_hom V W) : Prop :=
+∀ v, ∥f v∥ = ∥v∥
```

Comment:
Note that we have `isometry` in `mathlib`. And we have a lemma that says "an `add_monoid_hom` is an isometry iff it preserves the norm".

Reply:
thanks, refactored